### PR TITLE
New Adapter: Adview

### PR DESCRIPTION
**Pull request by AdviewOpen**
New Adapter: Adview
Docs PR: https://github.com/prebid/prebid.github.io/pull/3251

### CODE REVIEW

File: adapters/adview/adview.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adViewAdapter struct {
+	EndpointTemplate template.Template
+}
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:19]: Please rename this struct to simply `adapter`. There is no need to repeat the package name. There is no need to export it using an upper case letter. I recommend also changing `EndpointTemplate` to `endpointTemplate` as there is no need to export it using an upper case letter.

File: adapters/adview/adview.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+// Builder builds a new instance of the Adf adapter for the given bidder with the given config.
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:21]: Looks like you've got a copy/paste typo here. Please replace "Adf" with "AdView".

File: adapters/adview/adview.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:23]: Nitpick: Consider removing this empty line.

File: adapters/adview/adview.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	urlTemplate, err := template.New("endpointTemplate").Parse(config.Endpoint)
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:24]: Nitpick: Consider using a consistent name for the template. Please choose use either `endpointTemplate` or `urlTemplate` and use the name consistently.

File: adapters/adview/adview.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		reqData := &adapters.RequestData{
+			Method: http.MethodPost,
+			Body:   requestJSON,
+			Uri:    url,
+		}
+
+		requestData = append(requestData, reqData)
+	}
+
+	//return []*adapters.RequestData{requestData}, errors
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:94]: Please remove this comment.

File: adapters/adview/adview.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	var errors []error
+	var validImps = make([]openrtb2.Imp, 0, len(request.Imp))
+	requestData := make([]*adapters.RequestData, 0, len(request.Imp))
+
+	for _, imp := range request.Imp {
+
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errors = append(errors, &errortypes.BadInput{
+				Message: err.Error(),
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:45]: Consider using a more specific error message here and on line 53. Perhaps something along the lines of:
```
errors = append(errors, &errortypes.BadInput{
  Message: fmt.Sprintf("invalid imp.ext, %s", err),
})
```
```
errors = append(errors, &errortypes.BadInput{
  Message: fmt.Sprintf("invalid imp.ext bidder object, %s", err),
})
```

File: adapters/adview/adview.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		}
+
+		imp.TagID = advImpExt.MasterTagID //tagid means posid
+
+		//for adview bid request
+		if imp.Banner != nil {
+			if len(imp.Banner.Format) != 0 {
+				imp.Banner.H = &imp.Banner.Format[0].H
+				imp.Banner.W = &imp.Banner.Format[0].W
+			}
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:65]: `imp.Banner` is shared memory. You need to make a copy of the banner object before modifying it:
```
if imp.Banner != nil {
  if len(imp.Banner.Format) != 0 {
    bannerCopy := *imp.Banner
    bannerCopy.H = &imp.Banner.Format[0].H
    bannerCopy.W = &imp.Banner.Format[0].W
    imp.Banner = &bannerCopy
  }
}
```

File: adapters/adview/adview.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+		}
+
+		validImps = append(validImps, imp)
+		request.Imp = validImps
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:69]: This doesn't look right.

Do you intend for each request to contain just a single impression (aka, impression splitting)? If so, please remove the `validImps` slice and replace this line with `request.Imp = []openrtb2.Imp{imp}`.

As the code is now, each iteration of the loop will have the current `imp` and all previous imps due to appending the `imp` to `validImps` which is defined outside of the loop.

Please add a json test with multiple imps (2 will be fine) to ensure proper behavior of the impression splitting.

File: adapters/adview/adview.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+			})
+		}
+	}
+
+	return bidResponse, errors
+}
+
+// Builds endpoint url based on adapter-specific pub settings from imp.ext
+func (adapter *adViewAdapter) buildEndpointURL(params *openrtb_ext.ExtImpAdview) (string, error) {
+
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:145]: Nitpick: Consider removing this empty line.

File: adapters/adview/adview.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+		if imp.ID == impID {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo, nil
+			} else if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			}
+		}
+	}
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:162]: This method of determining the bid type is an anti-pattern. You see it in many adapters, but it's not a good solution. Does your bidding server return the bid type in an ext?

File: static/bidder-params/adview.json [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "title": "Adview Adapter Params",
+  "description": "A schema which validates params accepted by the adview adapter",
+  "type": "object",
+  "properties": {
+    "placementId": {
+      "type": "string",
+      "description": "An ID which identifies the placement selling the impression"
```
Comment (SyntaxNode) [inline static/bidder-params/adview.json:9]: Does it make sense to also include a minLength of 1 for the placementId?

File: adapters/adview/adview.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adViewAdapter struct {
+	EndpointTemplate template.Template
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:18]: The type needs to be `*template.Template`. This was changed recently. Due to this, the code in this PR does not compile.

File: adapters/adview/adview.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+			})
+			continue
+		}
+
+		imp.TagID = advImpExt.MasterTagID //tagid means posid
+
+		//for adview bid request
+		if imp.Banner != nil {
+			if len(imp.Banner.Format) != 0 {
+				imp.Banner.H = &imp.Banner.Format[0].H
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:63]: Please add a json test for the `len(imp.Banner.Format) != 0` condition.

File: adapters/adview/adviewtest/exemplary/banner-app.json [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+  "httpCalls": [
+    {
+      "expectedRequest": {
+        "uri": "https://bid.adview.com/agent/thirdAdxService/accountid01",
+        "body": {
+          "id": "some-request-id",
+          "device": {
+            "ua": "test-user-agent",
+            "ip": "123.123.123.123",
+	    "ifa":"00000000-0000-0000-0000-000000000000",
```
Comment (SyntaxNode) [inline adapters/adview/adviewtest/exemplary/banner-app.json:52]: Please fix the inconsistent whitespace in this line. It starts with a tab, but the rest of the file is using spaces.

File: adapters/adview/adviewtest/exemplary/native-app.json [hunk 15]
```diff
@@ -0,0 +1,144 @@
+{
+    "mockBidRequest": {
+      "id": "some-request-id",
+      "device": {
+        "ua": "test-user-agent",
+        "ip": "123.123.123.123",
+	"ifa":"00000000-0000-0000-0000-000000000000",
```
Comment (SyntaxNode) [inline adapters/adview/adviewtest/exemplary/native-app.json:7]: Please fix the inconsistent whitespace in this line.

File: adapters/adview/adviewtest/exemplary/native-app.json [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    "httpCalls": [
+      {
+        "expectedRequest": {
+          "uri": "https://bid.adview.com/agent/thirdAdxService/accountid01",
+          "body": {
+            "id": "some-request-id",
+            "device": {
+              "ua": "test-user-agent",
+              "ip": "123.123.123.123",
+	      "ifa":"00000000-0000-0000-0000-000000000000",
```
Comment (SyntaxNode) [inline adapters/adview/adviewtest/exemplary/native-app.json:53]: Please fix the inconsistent whitespace in this line.

File: adapters/adview/adviewtest/exemplary/video-app.json [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+  "httpCalls": [
+    {
+      "expectedRequest": {
+        "uri": "https://bid.adview.com/agent/thirdAdxService/accountid01",
+        "body": {
+          "id": "some-request-id",
+          "device": {
+            "ua": "test-user-agent",
+            "ip": "123.123.123.123",
+	    "ifa":"00000000-0000-0000-0000-000000000000",
```
Comment (SyntaxNode) [inline adapters/adview/adviewtest/exemplary/video-app.json:58]: Please fix the inconsistent whitespace in this line.

File: adapters/adview/adviewtest/params/race/banner.json [hunk 18]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (SyntaxNode) [inline adapters/adview/adviewtest/params/race/banner.json:1]: The params files are no longer necessary. We apologize, as the documentation update is still in progress. Please delete `/params/race/banner.json`, `/params/race/native.json`, and `/params/race/video.json`.

File: adapters/adview/params_test.go [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+var invalidParams = []string{
+	``,
+	`null`,
+	`true`,
+	`5`,
+	`4.2`,
+	`[]`,
+	`{}`,
```
Comment (SyntaxNode) [inline adapters/adview/params_test.go:47]: There is no need to include as many invalid type cases. I recommend leaving in ", "null", "{}" and removing "true", "5", "4.2", "[]".

File: adapters/adview/params_test.go [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+	`{"adCode": "string", "seatCode": 5, "originalPublisherid": "string"}`,
```
Comment (SyntaxNode) [inline adapters/adview/params_test.go:48]: There is no need to include this test case.

File: adapters/adview/params_test.go [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+	`{ "accountId": "accountid01" }`,
+	`{ "placementId": "", "accountId": "" }`,
```
Comment (SyntaxNode) [inline adapters/adview/params_test.go:50]: Please also add a test case where only placementId is provided.

File: exchange/adapter_builders.go [hunk 22]
```diff
@@ -150,6 +151,7 @@ func newAdapterBuilders() map[openrtb_ext.BidderName]adapters.Builder {
 		openrtb_ext.BidderAdtarget:          adtarget.Builder,
 		openrtb_ext.BidderAdtelligent:       adtelligent.Builder,
 		openrtb_ext.BidderAdvangelists:      advangelists.Builder,
+		openrtb_ext.BidderAdview:    	     adview.Builder,
```
Comment (SyntaxNode) [inline exchange/adapter_builders.go:154]: It looks like your company name is "AdView". You are welcomed to rename `BidderAdview` to `BidderAdView` to better match your brand.

File: openrtb_ext/imp_adview.go [hunk 23]
```diff
@@ -0,0 +1,6 @@
+package openrtb_ext
+
+type ExtImpAdview struct {
```
Comment (SyntaxNode) [inline openrtb_ext/imp_adview.go:3]: It looks like your company name is "AdView". You are welcomed to rename `ExtImpAdview` to `ExtImpAdView` to better match your brand.

**Review by SyntaxNode — COMMENTED** [review]
Thank you for your submission. I have a few functional comments and several style nitpicks. I recommend first focusing on the compilation issue and test failures (due to modifying shared banner memory).

File: adapters/adview/adview.go (hunk 2 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:21]: ok, fixed.

File: adapters/adview/adview.go (hunk 3 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:23]: thx & fixed.

File: adapters/adview/adview.go (hunk 4 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:24]: fixed.

File: adapters/adview/adview.go (hunk 5 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:94]: done.

File: adapters/adview/adview.go (hunk 6 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:45]: fixed.

File: adapters/adview/adview.go (hunk 7 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:65]: fixed.

File: adapters/adview/adview.go (hunk 9 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:145]: ok.

File: static/bidder-params/adview.json (hunk 11 above)
Comment (AdviewOpen) [reply static/bidder-params/adview.json:9]: yes, added.

File: adapters/adview/adview.go (hunk 12 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:18]: roger.

File: adapters/adview/adviewtest/exemplary/banner-app.json (hunk 14 above)
Comment (AdviewOpen) [reply adapters/adview/adviewtest/exemplary/banner-app.json:52]: done.

File: adapters/adview/adviewtest/exemplary/native-app.json (hunk 15 above)
Comment (AdviewOpen) [reply adapters/adview/adviewtest/exemplary/native-app.json:7]: fixed.

File: adapters/adview/adviewtest/exemplary/native-app.json (hunk 16 above)
Comment (AdviewOpen) [reply adapters/adview/adviewtest/exemplary/native-app.json:53]: done.

File: adapters/adview/adviewtest/exemplary/video-app.json (hunk 17 above)
Comment (AdviewOpen) [reply adapters/adview/adviewtest/exemplary/video-app.json:58]: ok.

File: adapters/adview/adviewtest/params/race/banner.json (hunk 18 above)
Comment (AdviewOpen) [reply adapters/adview/adviewtest/params/race/banner.json:1]: ok, removed.

File: adapters/adview/params_test.go (hunk 19 above)
Comment (AdviewOpen) [reply adapters/adview/params_test.go:47]: got it.

File: adapters/adview/params_test.go (hunk 20 above)
Comment (AdviewOpen) [reply adapters/adview/params_test.go:48]: removed now.

File: openrtb_ext/imp_adview.go (hunk 23 above)
Comment (AdviewOpen) [reply openrtb_ext/imp_adview.go:3]: Thx for your suggestion, we have changed it.

File: adapters/adview/adview.go (hunk 8 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:69]: just fixed the issue.

File: adapters/adview/adview.go (hunk 10 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:142]: thanks and i have changed the codes, but it seems also old logic (i see other adapters also do), so maybe you can show me the better way to solve this issue? Thx again.

File: adapters/adview/adview.go (hunk 13 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go]: ok, i added the banner-format.json test case.

**Comment by AdviewOpen** [comment]
Thanks for reviewing & comments, i had committed the codes for fixed.

File: adapters/adview/adview.go [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errors []error
+	var bidderExt adapters.ExtImpBidder
+	imp := &request.Imp[0]
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		errors = append(errors, &errortypes.BadInput{
+			Message: fmt.Sprintf("invalid imp.ext, %s", err),
+		})
+		return nil, append(errors, err)
```
Comment (bsardo) [inline adapters/adview/adview.go:42]: It looks like we are adding the error to `errors` twice here. I think this should be something like:
```
return nil, []error{&errortypes.BadInput{
	Message: fmt.Sprintf("invalid imp.ext, %s", err.Error()),
}}
```

File: adapters/adview/adview.go [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	//use adview
+	var advImpExt openrtb_ext.ExtImpAdView
+	if err := json.Unmarshal(bidderExt.Bidder, &advImpExt); err != nil {
+		errors = append(errors, &errortypes.BadInput{
+			Message: fmt.Sprintf("invalid bidderExt.Bidder, %s", err),
+		})
+		return nil, append(errors, err)
```
Comment (bsardo) [inline adapters/adview/adview.go:50]: Same comment as above. It looks like we are adding the error to `errors` twice here.

File: adapters/adview/adview.go (hunk 24 above)
Comment (SyntaxNode) [reply adapters/adview/adview.go:42]: nice catch. i missed this.

File: adapters/adview/adview.go [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+
+	bidder := &adapter{
+		endpoint: endpointTemplate,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var bidderExt adapters.ExtImpBidder
+	imp := &request.Imp[0]
```
Comment (SyntaxNode) [inline adapters/adview/adview.go:36]: I didn't notice this before. It looks like you're only processing the first impression. What happens if there are more than one impression? You may wish to consider imp splitting (see https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#makerequests) or make it clear in your docs that all other impressions will be ignored by this adapter.

**Review by SyntaxNode — COMMENTED** [review]
Thank for quickly responding to our feedback. I left one further comment, but please open a docs pr (see: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation) as that is a requirement for merging.

**Comment by bsardo** [comment]
This is looking good. I'm just waiting on the response to @SyntaxNode's comment on imp splitting.

File: adapters/adview/adview.go (hunk 26 above)
Comment (AdviewOpen) [reply adapters/adview/adview.go:36]: yes, our bid server can not support multi imps, we just process the first one.

**Comment by AdviewOpen** [comment]
OK，thanks and i have committed PR#3251 for user doc.

**Review by bsardo — APPROVED** [review]
LGTM

**Review by SyntaxNode — APPROVED** [review]

**Pull request merged**